ethology/tap_models/track_any_point.py

```diff
@@ -0,0 +1,163 @@
+                raise FileNotFoundError(
+                    f"Tracker model weights {weight_file} not found"
+                )
+
+            if weight_file is not None:
+                self.model = CoTrackerPredictor(checkpoint=weight_file)
+            else:
+                self.model = torch.hub.load(
+                    "facebookresearch/co-tracker", "cotracker3_offline"
+                )
+
+    def convert_to_movement_dataset(self, pred_tracks):
+        """Convert the predicted tracks to movement dataset.
+
+        Parameters
+        ----------
+        pred_tracks: torch.Tensor
+            Tensor containing tracks of each query
+            point across the frames of size
+            [batch, frame, query_points, X, Y].
+
+        Returns
+        -------
+        momentum.ValidPosesDataset
+            Dataset containing the predicted tracks.
+
+        """
+        # pred_tracks.shape =  [batch, frames, querypoints, 2d points]
+        # goes into positional array with shape
+        # [frames, 2d points, 1 keypoint per query point, no. of query points]
+        pred_tracks = pred_tracks.cpu().numpy().squeeze(0)
+        ds = from_numpy(
+            position_array=pred_tracks.transpose(0, 2, 1)[:, :, None, :],
+            source_software="cotracker",
+        )
+        return ds
+
+    def track(
+        self,
+        video_path: str,
+        query_points: list[list],
+        save_dir: str = "processed_videos",
+        save_results: bool = False,
+    ) -> torch.Tensor:
+        """Track the query points in the video source.
+
+        Parameters
+        ----------
+        video_path: str
+            Path to the input video file.
+        query_points: list
+            2D List of shape (Q,3) where Q is the
+            number of query points and each Q containing
+            Frame-Number to start tracking from, X, Y
+        save_dir: str
+            Directory path to save the processed video
+        save_results: bool
+            Flag to save the results as a video file
+
+        Returns
+        -------
+        torch.Tensor
+            Tensor containing tracks of each query
+            point across the frames of size
+            [batch, frame, query_points, X, Y].
+            Example: 50 frames video with 4 query_points
+            will give a tensor of shape [1, 50, 4, 2]
+
+        """
+        # check if video source file exists in path
+        if not os.path.isfile(video_path):
+            raise FileNotFoundError(
+                f"Video source file {video_path} not found"
+            )
+
+        # load video
+        video = read_video_from_path(video_path)
+
+        # (Frames, Height, Width, Channels) ->
+        # (Batch, Frames, Channels, Height, Width)
+        video = torch.from_numpy(video).permute(0, 3, 1, 2)[None].float()
+
+        # List to tensor
+        query_tensor = torch.tensor(query_points, dtype=torch.float32)
+
+        # set device
+        device = (
```

Comment:
Could we include `mps` if available, before considering `cpu`?

Reply:
also, when trying to run a larger video. I seem to run out of GPU memory (8 GB). Should the device be taken from user or based on the system itself? current priority cuda > mps > cpu

Reply:
We could allow the user to set it (via config?), but if not set you could have a default priority (cuda > mps > cpu)